Unordering receiver

martinthomson opened the pull request:
OK, the next small increment in my journey to smooth the flow of data between protocol and network.

This one takes a stream of data off the network and lets callers receive at different indices in any order.

You might have 10 messages, with receivers 0 through 9 running in different tasks.  Each can call `recv()` with their index (i.e., `RecordId`) in any order, from any thread.  The data will be read off as it arrives and delivered to each in order.

The form of the `Stream` that this consumes isn't *exactly* the same as the one that hyper produces from its requests, but a simple adapter should work for that.

This isn't the whole of the work involved, but it's the fiddly part.  From here, there is a bunch of plumbing to go.  I might be able to split off the code for matching incoming streams to steps, but the rest might have to go all together, because the integration starts to get harder to do in parts.

akoshelev commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+        }
+        // See the documentation for `overflow_wakers`.
+        if self.next % (self.wakers.len() / 2) == 0 {
+            for w in self.overflow_wakers.drain(..) {
```
I think `drain` never shrinks the allocated capacity so the memory footprint may be less than ideal. Is it reasonable to assume the average capacity to be equal to `self.wakers.capacity()` and set it after each drain?

martinthomson replied:
Yeah, I might instead use `std::mem::take()` here.

akoshelev commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+use typenum::Unsigned;
+
+/// A future for receiving item `i` from an `UnorderedReceiver`.
+#[pin_project]
```
It feels that you don't need to pin anything. `usize` is `Unpin` and `OperatingState` pins the stream

martinthomson replied:
Hmm, I added this early on and didn't come back to it (what you see is not what I started with).  I'll take another look.

akoshelev commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+                "Awaiting a read that has already been fulfilled"
+            );
+            let index = i % self.wakers.len();
+            self.wakers[index] = Some(waker);
```
maybe another debug assert that `self.wakers[index] == None`?

martinthomson replied:
That's not going to be the case.  I originally had that and found that there are some cases (the synchronous test is one), where we replace a waker.  Basically, if the future is polled more than once in this state, we will be getting another waker for the same item.  I think that a strict interpretation of the `Future` contract requires us to store both wakers in that case, but the note above explains why I think that's unnecessary.

(I spent a lot of time thinking about this and I'm somewhat sure that this is not OK in theory, but maybe OK in practice.  That is, what if the first to poll is kept, the second is polled and discarded? We won't wake the first.  But then, we won't be doing that, I hope.)

martinthomson replied:
I've got a check for this that should ensure that it is robust.  Not guaranteed (`Waker::will_wake` can return false in cases where there is no issue), but then we shouldn't be hitting that code path outside of the few tests I have.

akoshelev commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+    /// So we don't want to wake them frequently.  Instead, these are woken on
+    /// a fixed cadence of every `c/2` items.
+    ///
+    /// When polled again, any that were only a little past the capacity will
```
is it worth considering a future optimization where we only poll overflow wakers when there is nothing to be polled in `wakers`?

martinthomson replied:
That might be better, the more I think on it.  I was concerned that we might have to walk the main store to ensure that things were fine, but we can count the number of wakers that are in there.  The other cost is that we might flush the extras too often, but I think that only happens after you finish reading the last item in a chunk in most cases.  With that design, any time you read an item out and there are no outstanding receivers, you will flush the overflow.  That will flush more often than this, probably, but it might be simpler to reason about.

akoshelev replied:
> I was concerned that we might have to walk the main store to ensure that things were fine, but we can count the number of wakers that are in there.

yea I was also thinking about maintaining a counter of active wakers inside `wakers`. that shouldn't require walking through the vector and counting

akoshelev reviewed (commented):
very elegant solution!

martinthomson commented on src/bits/mod.rs:
```diff
@@ -71,5 +71,5 @@ pub trait Serializable: Sized {
     ///
     /// [`serialize`]: Self::serialize
-    fn deserialize(buf: GenericArray<u8, Self::Size>) -> Self;
+    fn deserialize(buf: &GenericArray<u8, Self::Size>) -> Self;
```
We had a few cases where we used `GenericArray::clone_from_slice` in order to produce this value.  That involved a copy, which wasn't necessary, whereas `GenericArray::from_slice` doesn't.  But that returns a reference.

akoshelev replied:
that's right. I don't mind changing it this way, just saying that for places where I believe currently deserialize is "free" (bitarray) it will come with a cost of cloning.

martinthomson replied:
Yeah, though I realized that this is what we really want in those cases.  We want to own the copy of whatever came off the wire.  I also discovered that those cases can rely on the type being `Copy`, so it's cheap.

martinthomson commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+    /// [`poll`]: Future::poll
+    fn add_waker(&mut self, i: usize, waker: Waker) {
+        // We don't save a waker at `self.next`, so `>` and not `>=`.
+        if i > self.next + self.wakers.len() {
```
My understanding is that capacity is a measure of the size of the underlying store, which could be larger than what you asked for.  If the allocation is larger, then this code will be incorrect.

akoshelev replied:
right, I thought if len() != capacity() this code will be either incorrect or super inefficient, potentially storing everything but one item inside the overflow wakers.

martinthomson commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+            w.wake();
+        }
+        // See the documentation for `overflow_wakers`.
+        if self.next % (self.wakers.len() / 2) == 0 {
```
Well, *this* design depends on this being done unconditionally, but I think I'm going to try your suggested design instead.

akoshelev replied:
ah I think I got it now. You wake the overflow wakers not just because there might be a chance that their future is ready, but also because you want to move some of them to the `wakers` vector. In this case you have to wake them up on schedule

martinthomson commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,502 @@
+    /// this should be the same as the value given to [`ordering_mpsc`].
+    ///
+    /// [`ordering_mpsc`]: crate::helpers::buffers::ordering_mpsc::ordering_mpsc
+    pub fn new(stream: Pin<Box<S>>, capacity: NonZeroUsize) -> Self {
```
I don't think that makes this easier in this case.  It would make it less flexible for one (this allows you to read different-sized messages).  This basically takes that functionality and implements it internally, which should be more efficient overall.

martinthomson commented on src/bits/bit_array.rs:
```diff
@@ -10,8 +10,17 @@ type U8_1 = BitArr!(for 8, in u8, Lsb0);
+
 macro_rules! bit_array_impl {
     ( $modname:ident, $name:ident, $store:ty, $bits:expr, $arraylen:ty ) => {
+        // $store
+
```
```suggestion
```

martinthomson commented:
OK, I just implemented a change to track wakers differently and hit a deadlock.

The logic I implemented looks like this:

```rust
if let Some(w) = self.wakers[self.next % N].take() {
  w.wake();
  self.active_wakers -= 1;
} else if self.active_wakers == 0 {
  for w in std::mem::take(&mut self.overflow_wakers) { 
    w.wake();
  }
}
```

That is, if we ever find that there are no wakers left in the active set, flush the overflow.  By reactivating those, you will get more things to work on.

The problem that leads to a deadlock is where you get a gap.

Let's say that you wake item N, which is the last item in the active store.  Item N+1 is in the overflow, but you know that item N is coming back soon, so you don't bother flushing out the store just yet.

Before item N is polled again, item N+2 is polled.  This is put into the active store.  Now, when item N is polled again, the store is not empty, but item N+1 is stuck in the overflow.

I conclude from this that you have to flush the overflow any time that it is not empty and the next slot in the active store is empty.  That's much less efficient.

I'm going to go back to the c/2 design I started out with :)  Thanks to shuttle again for finding this one (I'm also glad I used a capacity of 3 rather than a capacity of 2, because this only happens for c>2).

akoshelev commented on src/helpers/buffers/unordered_receiver.rs:
```diff
@@ -0,0 +1,504 @@
+    type Output = Result<M, Error>;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let this = self.as_ref();
```
you could use `self` because pin implements `Deref`
